### DISCUSSION TITLE
How to fix compressed Aggregates response?

### DISCUSSION THREAD

**Discussion by LukasSefcik** [ThreeMammals/Ocelot#1900] [Q&A] [answered]
## Expected Behavior
![ocelot_aggregation_encoding](https://user-images.githubusercontent.com/24602626/49872926-d3c04c80-fe1a-11e8-9996-518c776bb568.png)

## Actual Behavior

When using request aggregation we are getting response which looks like it's compressed two times.
First time is compressed whole response and second time is compressed content of each endpoint response.

When added `Accept-Encoding: br` or `Accept-Encoding: utf-8` to request header everything works perfectly. 

Using Angular 7 HttpClient We cannot set Accept-Encoding because of 'Refused to set unsafe header "Accept-Encoding"'.

Any suggestions, please?

Thank you

![ocelot_aggregation](https://user-images.githubusercontent.com/24602626/49872715-3bc26300-fe1a-11e8-85a1-f33b21ebaf1c.png)

## Specifications

  - Version: 12.0.1
  - Platform: .NET Core 2.1
  - Subsystem:

**Comment by ramesh26** [comment]
@LukasSefcik @philproctor ,

Hi, I've also faced the same issue which I reported below. Can you please provide solution for this issue

https://github.com/ThreeMammals/Ocelot/issues/1072

**Comment by MSeifAllah** [comment, marked as answer]
@LukasSefcik @philproctor ,
I had the problem . You have the problem because the responses from Routes are 
compressed (  Content-Encoding : gzip ) 

to fixed it you have two solutions  : 

-  add a delegate to your aggregator and add 
```csharp
protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
{
    request.Headers.AcceptEncoding.Clear();
    request.Headers.AcceptEncoding.Add(new StringWithQualityHeaderValue("deflate"));
    request.Headers.AcceptEncoding.Add( new StringWithQualityHeaderValue("utf-8"));
    request.Headers.AcceptEncoding.Add(new StringWithQualityHeaderValue("br"));
    return await base.SendAsync(request, cancellationToken);
}
```
- implements `IDefinedAggregator` and decompress the responses.

**Reply by raman-m** [reply]
Why not to create a PR to deal with this compression gotcha?

**Comment by fadem** [comment]
hi, I have a question about the request to aggregates from Angular CLI.
I set the configuration that I needed for "Aggregates". Normally this Route work on the requested URL but when I try to this on Angular CLI it was returned response "Cors Origin Error". I set the configuration for "Cors Origin Error" and those are work one by one without returned response "Cors Origin Error".

**Comment by neetra** [comment]
@fadem  @LukasSefcik  @ramesh26  
based on @MSeifAllah  suggestion
[here](https://github.com/ThreeMammals/Ocelot/issues/1207#issuecomment-618895233) example how to solve this issue

**Comment by natiox** [comment]
Here's more detail on solution 1 from @MSeifAllah.  

That solution is talking about a [Delegating Handler](https://ocelot.readthedocs.io/en/latest/features/delegatinghandlers.html) in Ocelot.  I preferred that because it didn't involve a bunch of custom C# code for each aggregate request.  However, I don't know what the side effects might be as far as replacing gzip compression with other types of compression.  Seems to work for my simple test, though.

```C#
using System.Net.Http;
using System.Net.Http.Headers;
using System.Threading;
using System.Threading.Tasks;
...
public class NoGzipDelegatingHandler : DelegatingHandler
{        
    protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
    {
        //Trying to avoid gzip encoding in the response, so selecting other compression algorithms
        request.Headers.AcceptEncoding.Clear();
        request.Headers.AcceptEncoding.Add(new StringWithQualityHeaderValue("deflate"));
        request.Headers.AcceptEncoding.Add(new StringWithQualityHeaderValue("br"));
        return await base.SendAsync(request, cancellationToken);
    }
}
```

Then in Startup.cs, you add your delegating handler as follows.  (The boolean parameter on `AddDelegatingHandler` set to true makes the handler global across all requests.)
```C#
public void ConfigureServices(IServiceCollection services)
{
    services.AddOcelot(Configuration)
    .AddDelegatingHandler<NoGzipDelegatingHandler>(true);
}
```

**Reply by raman-m** [reply]
Why not to create a PR to deal with this compression gotcha?
